crd/src/util.rs

```diff
@@ -0,0 +1,674 @@
+///
+/// * `opa_spec` - OpaSpec to retrieve different port configuration
+/// * `opa_pods` - Pods belonging to the OPA cluster
+/// * `opa_api` - The desired OPA endpoint like Data, Query, Policy etc.
+/// * `opa_api_protocol` - The desired OPA endpoint protocol like HTTP or HTTPS
+/// * `desired_node_name` - If node_name is provided we look for opa deployments on the same node name to improve response time
+///
+fn get_opa_connection_string_from_pods(
+    opa_spec: &OpaSpec,
+    opa_pods: &[Pod],
+    opa_api: &OpaApi,
+    opa_api_protocol: &OpaApiProtocol,
+    desired_node_name: Option<String>,
+) -> OpaOperatorResult<String> {
+    let mut server_and_port_list = Vec::new();
+
+    for pod in opa_pods {
+        let pod_name = match pod.metadata.name.as_ref() {
+            None => {
+                return Err(ObjectWithoutName {
+                    reference: ErrOpaPodWithoutName.to_string(),
+                })
+            }
+            Some(pod_name) => pod_name,
+        };
+
+        let node_name = match pod.spec.as_ref().and_then(|spec| spec.node_name.as_ref()) {
+            None => {
+                debug!("Pod [{:?}] is does not have node_name set, might not be scheduled yet, aborting.. ",
+                       pod_name);
+                return Err(PodWithoutHostname {
+                    pod: pod_name.clone(),
+                });
+            }
+            Some(node_name) => node_name,
+        };
+
+        let role_group = match pod
+            .metadata
+            .clone()
+            .labels
+            .unwrap_or_default()
+            .get(APP_ROLE_GROUP_LABEL)
+        {
+            None => {
+                return Err(PodMissingLabels {
+                    labels: vec![String::from(APP_ROLE_GROUP_LABEL)],
+                    pod: pod_name.clone(),
+                })
+            }
+            Some(role_group) => role_group.to_owned(),
+        };
+
+        let opa_port = get_opa_port(&opa_spec, &role_group)?;
+
+        // if a node_name is provided we prefer OPA deployments that are located on the same machine
+        if let Some(desired_host) = &desired_node_name {
+            if node_name == desired_host {
+                let url = opa_api.get_url(opa_api_protocol, &node_name, opa_port)?;
+                debug!(
+                    "Found Opa deployment on provided node [{}]; Using this one [{}] ...",
+                    node_name, url
+                );
+                return Ok(url);
+            }
+        }
+
+        server_and_port_list.push((node_name, opa_port));
+    }
+
+    // Sort list by hostname to make resulting connection strings predictable
+    // Shouldn't matter for connectivity but makes testing easier and avoids unnecessary
+    // changes to the infrastructure
+    server_and_port_list.sort_by(|(host1, _), (host2, _)| host1.cmp(host2));
+
+    // TODO: randomly search for node? -> for now we just take the last one
+    let server = server_and_port_list.pop();
```

Comment:
Yeah, this is not great for load balancing purposes.
Could you pick a random one?

https://stackoverflow.com/questions/34215280/how-can-i-randomly-select-one-element-from-a-vector-or-array/34215930

seems relatively simple